File: netlify/functions/spotify-search.js
// Netlify Serverless Function for Spotify Search
const { loadEnv } = require('./utils/load-env'); // Load environment variables from appropriate .env file
loadEnv();

const { createClient } = require('@supabase/supabase-js');
const { headers } = require('./utils/cors-headers');
const fetch = require('node-fetch');

exports.handler = async (event, context) => {
  // Only allow GET requests
  if (event.httpMethod !== 'GET') {
    return {
      statusCode: 405,
      headers,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  }

  // Parse query parameters
  const params = new URLSearchParams(event.queryStringParameters);
  const query = params.get('query');
  const limit = params.get('limit') || 10;

  if (!query) {
    return {
      statusCode: 400,
      headers,
      body: JSON.stringify({ error: 'Missing query parameter' })
    };
  }

  // Get Spotify client ID and secret from environment variables
  const clientId = process.env.SPOTIFY_CLIENT_ID;
  const clientSecret = process.env.SPOTIFY_CLIENT_SECRET;

  if (!clientId || !clientSecret) {
    console.error('Missing Spotify credentials');
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ 
        error: 'Server configuration error', 
        detail: 'Missing Spotify credentials' 
      })
    };
  }

  try {
    // Get an access token from Spotify
    const tokenResponse = await fetch('https://accounts.spotify.com/api/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': 'Basic ' + Buffer.from(clientId + ':' + clientSecret).toString('base64')
      },
      body: 'grant_type=client_credentials'
    });
    
    const tokenData = await tokenResponse.json();
    
    if (tokenData.error) {
      console.error('Error getting Spotify token:', tokenData);
      
      // Return more helpful error messages based on the error type
      if (tokenData.error === 'invalid_client') {
        return {
          statusCode: 500,
          headers,
          body: JSON.stringify({ 
            error: 'Failed to authenticate with Spotify',
            message: 'Invalid Spotify client credentials. Please see FIX_SPOTIFY_CREDENTIALS.md for instructions on how to fix this issue.',
            details: tokenData
          })
        };
      }
      
      return {
        statusCode: 500,
        headers,
        body: JSON.stringify({ 
          error: 'Failed to authenticate with Spotify',
          message: tokenData.error_description || 'Spotify API returned an error',
          details: tokenData
        })
      };
    }

    // Search Spotify using the token
    const spotifyResponse = await fetch(
      `https://api.spotify.com/v1/search?q=${encodeURIComponent(query)}&type=track&limit=${limit}`, {
        headers: {
          'Authorization': `Bearer ${tokenData.access_token}`
        }
      }
    );

    const spotifyData = await spotifyResponse.json();

    if (!spotifyResponse.ok) {
      console.error('Error from Spotify API:', spotifyData);
      return {
        statusCode: spotifyResponse.status,
        headers,
        body: JSON.stringify({ 
          error: 'Error from Spotify API', 
          message: spotifyData.error.message || 'Spotify API returned an error',
          details: spotifyData
        })
      };
    }

    // Process and return the response
    const tracks = spotifyData.tracks.items.map(track => ({
      id: track.id,
      name: track.name,
      uri: track.uri,
      artists: track.artists.map(artist => artist.name).join(', '),
      album: track.album.name,
      duration_ms: track.duration_ms,
      album_art: track.album.images[0]?.url || null
    }));

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({
        tracks,
        total: spotifyData.tracks.total
      })
    };
  } catch (error) {
    console.error('Error searching Spotify:', error);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ 
        error: 'An error occurred while searching Spotify',
        message: error.message,
        details: error
      })
    };
  }
};

File: netlify/functions/spotify-search/spotify-search.py
# netlify/functions/spotify-search/spotify-search.py
import sys
import os
import json

# Add the deps directory to the Python path
current_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.append(os.path.join(current_dir, "deps"))

# Now import your dependencies
import requests
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

def handler(event, context):
    """
    Netlify Function handler for Spotify search
    """
    try:
        # Parse the request body
        body = json.loads(event['body']) if event.get('body') else {}
        query = body.get('query', '')
        
        if not query:
            return {
                'statusCode': 400,
                'body': json.dumps({'error': 'Missing search query'})
            }
        
        # Here you would implement the Spotify search logic
        # For example, fetch from Spotify API using requests
        
        # For now, return a mock response
        return {
            'statusCode': 200,
            'body': json.dumps({
                'success': True,
                'data': {
                    'tracks': [
                        {'id': '1', 'name': 'Example Track 1', 'artist': 'Artist 1'},
                        {'id': '2', 'name': 'Example Track 2', 'artist': 'Artist 2'}
                    ]
                }
            })
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'body': json.dumps({'error': str(e)})
        }
